### ISSUE TITLE
Legacy 1.x 32-bit Atomic Extension: Volatile or not

### ISSUE THREAD

**Issue by alycm** [KhronosGroup/OpenCL-Docs#322]
This was internal issue 219.

The OpenCL 1.0 32-bit atomic extensions that became Core in 1.1 have the form `int atom_add (__global int *p, int val)`. The move to Core renamed them to `atomic_add` etc, and added `volatile`.  However, the original `atom_add` names are were always still explicitly supported.

When documenting this in OpenCL C 3.0 provisional the original names also had `volatile` added to their definition.  All the open-source implementations that I checked actually have `volatile` here (i.e. contrary to previous versions of the spec).

So if the bug that OpenCL 3.0 added `volatile` where it was not previously, or is the bug that `volatile` was previously missing when it should have been specified.

**Comment by alycm** [comment]
One obvious difference to a programmer is that passing a volatile pointer to a non-volatile function may result in  a warning.  That doesn't happen due to adding `volatile` to these functions as has been done, it could happen to existing code if implementions remove `volatile` to match older specifications (or 3.0 if it matches those older specifications).

```c
int atom_nonvolatile(global int *p, int val) {
   return atom_add(p, val);
}

kernel void test(global int* A, volatile global int* B)
{
    atomic_nonvolatile(A, 1);
    atomic_nonvolatile(B, 1);  // May warn about discarding qualifiers.
}
```

**Comment by bashbaug** [comment]
The C11 atomic functions include `volatile` also.

Unless there's a strong reason NOT to include `volatile` I think we should call this a spec bug that is now fixed since the OpenCL C 3.0 spec includes `volatile`, and close this issue with no action required.

(Reminder: we also updated the old 32-bit atomics extension specs [here](https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#cl_khr_int32_atomics) as well.)

**Comment by bashbaug** [comment]
Closing this issue as discussed in the June 30th teleconference.
